### ISSUE TITLE
Question on forecasting method

### ISSUE THREAD

**Issue by jermaine1ronquillo** [PYFTS/pyFTS#6]
I'm new to fuzzy logic and I'd like know why the predict method requires the test set of the data?
Another question is does the method predict only one-step ahead?

Regards

**Comment by petroniocandido** [comment]
It is not exactly the "test" data, but the necessary lags used to forecast. On a first order model (as pyFTS.models.chen.ConventionalFTS) all you need to forecast t+1 is the last lag, i. e., a list with the last value of the time series. If you are using a high order model (pyFTS.models.hofts.HighOrderFTS) you will need more  past lags.

Not all methods are projected to work with multiple steps ahead forecasting, but on the FTS.predict method there is a parameter "steps" where you can indicate the forecasting horizon you need. In this case the method will feedback its outputs for each step ahead.

Don't hesitate to contact us if you have more questions!

Best regards.

**Comment by jermaine1ronquillo** [comment]
Thank you for the response, although I'm still a bit confused. I will try pyFTS to forecast water quality parameters,  our objectives are to forecast at least 3-day ahead of time and able to predict if the parameters will be above a certain threshold. Definitely I'll ask more questions regarding the use of pyFTS. Again thank you.

**Comment by petroniocandido** [comment]
Hi!

Check this out: https://colab.research.google.com/drive/1yeaYrgasByD12JI-nEIlE_buQft3YR3I 

The minimal input length for the predict method is the order of the model! To forecast multiple steps ahead you just need to use the parameter steps_ahead, indicating how many steps to forecast. 


About the water quality time series, is it seasonal? Mono or Multivariate? How looks its ACF?

**Comment by jermaine1ronquillo** [comment]
Thank you for  this, based on my initial exploration on the data the time series show small seasonality  (using statsmodels)
Below are  the graphs
![image](https://user-images.githubusercontent.com/15191328/45339079-48d57900-b5c3-11e8-8f17-64fb36a401dc.png)
![image](https://user-images.githubusercontent.com/15191328/45339091-57239500-b5c3-11e8-82eb-1380ce57c44a.png)

**Comment by petroniocandido** [comment]
It is a public dataset?

**Comment by jermaine1ronquillo** [comment]
Unfortunately it is not, this was recorded from our treatment facility. Are you interested with the data? Please provide me with your email address and maybe I might get an approval from my superiors. 

Regards

**Comment by jermaine1ronquillo** [comment]
Thank for your time and effort!
Here is my code, basically I copied your example.

Regards

train=data_mod['2012':'2016']['Turbidity'].values
test=data_mod['2017']['Turbidity'].values
from pyFTS.partitioners import Grid
fig, ax = plt.subplots(nrows=1, ncols=1,figsize=[25,5])
fs = Grid.GridPartitioner(data=train, npart=35)
fs.plot(ax)
from pyFTS.models import hofts
model1 = hofts.HighOrderFTS(order=1, partitioner=fs)
model1.fit(train)
print(model1)
fig, ax = plt.subplots(nrows=1, ncols=1,figsize=[15,5])
forecasts = model1.predict(test)
ax.plot(test[80:100], label='test')
ax.plot(forecasts[80:100], label='forecast')
ax.legend()
test_mod=data_mod['2017']['Turbidity'].copy()
test_mod['2017-04-04']=10
fig, ax = plt.subplots(nrows=1, ncols=1,figsize=[15,5])
forecasts = model1.predict(test_mod)
ax.plot(test_mod.values[80:100], label='test')
ax.plot(forecasts[80:100], label='forecast')
ax.legend()
test_mod['2017-04-05']=10
fig, ax = plt.subplots(nrows=1, ncols=1,figsize=[15,5])
forecasts = model1.predict(test_mod)
ax.plot(test_mod.values[80:100], label='test')
ax.plot(forecasts[80:100], label='forecast')
ax.legend()

**Comment by petroniocandido** [comment]
Looks fine to me! 

Try higher order models  to improve the accuracy.

Best regards!

**Comment by pintuiitbhi** [comment]
I trained a HOFTS with order 3 on erratic data. Dataset is montly. I predicted on train data and it looks like it has overfitted. 
![train](https://user-images.githubusercontent.com/29093550/59016087-0bd7a380-8830-11e9-97c5-0bd4dd697cad.png)

Here is the code:
```
from pyFTS.models import hofts, pwfts

fig, ax = plt.subplots(nrows=1, ncols=1, figsize=[15,8])


ax.plot(train_uv[:100], label='Original')
rows = []
# for method in [hofts.HighOrderFTS, hofts.WeightedHighOrderFTS, pwfts.ProbabilisticWeightedFTS]:
for method in [hofts.HighOrderFTS]:
#     for order in [1, 2,3]:
    for order in [3]:
        model = method(partitioner=part, order=order)

        model.shortname += str(order)

        model.fit(train_uv)

        forecasts = model.predict(train_uv)
#         forecasts = model.predict([0,792,492], steps_ahead=142)
        forecast_fuzzy = forecasts
        for k in np.arange(order):
            forecasts.insert(0,None)

        ax.plot(forecasts[:100], label=model.shortname)

        models.append(model.shortname)

#         Util.persist_obj(model, model.shortname)

#         del(model)

handles, labels = ax.get_legend_handles_labels()
lgd = ax.legend(handles, labels, loc=2, bbox_to_anchor=(1, 1))  

```

**I want to predict next 3 months data ie. 12 data points. How I can use the "predict" function to do this?**

This is how I tried.
```
fig, ax = plt.subplots(nrows=1, ncols=1, figsize=[15,8])

ax.plot(test_uv, label='Original')
forecasts = model.predict([1068,2280,4392], steps_ahead=12)
order = 3
for k in np.arange(order):
    forecasts.insert(0,None)
ax.plot(forecasts, label=model.shortname)
handles, labels = ax.get_legend_handles_labels()
lgd = ax.legend(handles, labels, loc=2, bbox_to_anchor=(1, 1))
```

![test](https://user-images.githubusercontent.com/29093550/59016410-b059e580-8830-11e9-8faa-9b3282f4e95c.png)

**Why is it remains constant after predicting 3 points in future?** 

> [1068,2280,4392] is the last three data points of train dataset.

**Comment by ramdhan1989** [comment]
> Can you share your code for verification? I'm working on a pyFTS tutorial for solar forecasting and the results are very good (texts in portuguese), around 5% of error (MAPE): https://colab.research.google.com/drive/1xfonrM853rtWTsVet7oJsFO-OoHWRgk6
> 
> The quality of a FTS model depends on several factors:
> a) Data quality (in general FTS models are very sensitive to outliers);
> b) Method (different methods for different demands);
> c) Transformations (pre and post processing operations)
> d) Partitioning (few partitions will underfit the model, to much partitions will overfit the model);
> e) Order (the minimal number of lags used by the model);
> f) Lags indexes (which past lags produce better generalizations);
> g) alpha_cut (the minimal membership grade considered on fuzzyfication step, it helps to reduce overfit by cutting useless rules)
> 
> The default values of the FTS methods generally fit well on data. But depending on you application domain it is necessary to perform a parameter fine tuning . This hyperparameter optimization can performed using a genetic algorithm (I like to use the DEAP library for evolutive optimization: https://github.com/DEAP/deap) or specific hyperparameter optimization library such as hyperopt (https://github.com/hyperopt/hyperopt).
> 
> I hope I have helped, but still in contact for any questions!

did you have example code to use hyperparam technique for pyFTS package ?

**Comment by petroniocandido** [comment]
Hi @ramdhan1989 !

Please check the issue https://github.com/PYFTS/pyFTS/issues/30
